hw/ip/otbn/code-snippets/otbn_bn_mul64.asm

```diff
@@ -0,0 +1,549 @@
+JALR x0, x1, 0
+
+mul1_exp:
+BN.LID x9, 0(x17)
+BN.MOV w2, w31
+LOOP x30, 1
+BN.MOVR x10++, x11
+BN.XOR w2, w2, w2
+BN.ADDI w2, w2, 1
+ADDI x6, x16, 0
+ADDI x7, x19, 0
+LOOP x30, 4
+ADDI x16, x6, 0
+ADDI x19, x7, 0
+JAL x1, mma
+BN.MOV w2, w31
+ADDI x16, x6, 0
+ADDI x19, x7, 0
+BN.SUB w2, w2, w2, FG1
+LOOP x30, 3
+BN.LID x9, 0(x16++)
+BN.MOVR x11, x8++
+BN.CMPB w3, w2, FG1
+ADDI x8, x0, 4
+ADDI x10, x0, 4
+ADDI x16, x6, 0
+ADDI x19, x7, 0
+JAL x1, mm1_sub_cx
+ADDI x16, x6, 0
+ADDI x19, x7, 0
+JALR x0, x1, 0
+
+mul1:
+ADDI x3, x0, 0
+BN.LID x3, 0(x0)
+JAL x1, setupPtrs
+JAL x1, mul1_exp
+JALR x0, x1, 0
+
+sqrx_exp:
+LW x16, 8(x0)
+LW x17, 9(x0)
+LW x18, 10(x0)
+LW x19, 11(x0)
+LW x20, 12(x0)
+LW x21, 13(x0)
+LW x22, 14(x0)
+LW x23, 15(x0)
+BN.LID x9, 0(x17)
+BN.MOV w2, w31
+LOOP x30, 1
+BN.MOVR x10++, x11
+LW x10, 2(x0)
+LOOP x30, 8
+BN.LID x11, 0(x20++)
+ADDI x5, x20, 0
+ADDI x6, x16, 0
+ADDI x7, x19, 0
+JAL x1, mma
+ADDI x20, x5, 0
+ADDI x16, x6, 0
+ADDI x19, x7, 0
+ADDI x10, x0, 4
+ADDI x8, x0, 4
+LOOP x30, 2
+BN.SID x8++, 0(x21)
+ADDI x21, x21, 1
+ADDI x8, x0, 4
+ADDI x10, x0, 4
+LW x12, 4(x0)
+LW x13, 5(x0)
+JALR x0, x1, 0
+
+mulx_exp:
+LW x16, 16(x0)
+LW x17, 17(x0)
+LW x18, 18(x0)
+LW x19, 19(x0)
+LW x20, 20(x0)
+LW x21, 21(x0)
+LW x22, 22(x0)
+LW x23, 23(x0)
+BN.LID x9, 0(x17)
+BN.MOV w2, w31
+LOOP x30, 1
+BN.MOVR x10++, x11
+ADDI x8, x0, 4
+ADDI x10, x0, 4
+LW x12, 4(x0)
+LW x13, 5(x0)
+LOOP x30, 8
+BN.LID x11, 0(x20++)
+ADDI x5, x20, 0
+ADDI x6, x16, 0
+ADDI x7, x19, 0
+JAL x1, mma
+ADDI x20, x5, 0
+ADDI x16, x6, 0
+ADDI x19, x7, 0
+ADDI x8, x0, 4
+ADDI x10, x0, 4
+LW x12, 4(x0)
+LW x13, 5(x0)
+JALR x0, x1, 0
+
+selOutOrC:
+BN.WSRRW w3, 2, w3
+BN.OR w3, w3, w3
+CSRRS x2, 1984, x0
+ANDI x2, x2, 2
+BNE x2, x0, selOutOrC_invsel
+BN.ADDC w3, w3, w3 << 16B
+LOOP x30, 10
+BN.WSRRW w3, 2, w3
+BN.WSRRW w2, 2, w2
+BN.LID x9, 0(x21)
+BN.SID x11, 0(x21)
+BN.MOVR x11, x8++
+BN.WSRRW w0, 2, w0
+BN.MOV w0, w2
+BN.WSRRW w2, 2, w2
+BN.SEL w2, w0, w3, L
+BN.SID x11, 0(x21++)
+JALR x0, x1, 0
+ADDI x0, x0, 0
+
+selOutOrC_invsel:
+BN.ADDC w3, w3, w3 << 16B
+LOOP x30, 10
+BN.WSRRW w3, 2, w3
+BN.WSRRW w2, 2, w2
+BN.LID x9, 0(x21)
+BN.SID x11, 0(x21)
+BN.MOVR x11, x8++
+BN.WSRRW w0, 2, w0
+BN.MOV w0, w2
+BN.WSRRW w2, 2, w2
+BN.SEL w2, w3, w0, L
+BN.SID x11, 0(x21++)
+JALR x0, x1, 0
+
+modexp:
+JAL x1, mulx
+LW x16, 24(x0)
+LW x17, 25(x0)
+LW x18, 26(x0)
+LW x19, 27(x0)
+LW x20, 28(x0)
+LW x21, 29(x0)
+LW x22, 30(x0)
+LW x23, 31(x0)
+BN.SUB w2, w2, w2
+LOOP x30, 4
+ADDI x0, x0, 0
+BN.LID x11, 0(x16++)
+BN.SUBB w2, w31, w2
+BN.SID x11, 0(x21++)
+ADDI x0, x0, 0
+SLLI x24, x22, 8
+LOOP x24, 19
+JAL x1, sqrx_exp
+JAL x1, mulx_exp
+LW x16, 24(x0)
+LW x17, 25(x0)
+LW x18, 26(x0)
+LW x19, 27(x0)
+LW x20, 28(x0)
+LW x21, 29(x0)
+LW x22, 30(x0)
+LW x23, 31(x0)
+BN.WSRRW w2, 2, w2
+BN.ADD w2, w2, w2
+LOOP x30, 4
+BN.WSRRW w2, 2, w2
+BN.LID x11, 0(x20)
+BN.ADDC w2, w2, w2
+BN.SID x11, 0(x20++)
+JAL x1, selOutOrC
+ADDI x0, x0, 0
+ADDI x3, x0, 0
+BN.LID x3, 48(x0)
```

Comment:
Shouldn't this be 96? (256/8*3)

Reply:
Erk. For some reason, I decided that WLEN was 16 bytes, not 32. Thanks for catching that!